pkg/backoff/backoff.go

```diff
@@ -0,0 +1,124 @@
+package backoff
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	ini "github.com/wakatime/wakatime-cli/pkg/config"
+	"github.com/wakatime/wakatime-cli/pkg/heartbeat"
+	"github.com/wakatime/wakatime-cli/pkg/log"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// resetAfter sets the total seconds a backoff will last.
+	resetAfter = 3600
+	// factor is the total seconds to be multiplied by.
+	factor = 15
+)
+
+// Config defines backoff data.
+type Config struct {
+	// At is the time when the first failure happened.
+	At *time.Time
+	// Retries is the number of attempts to connect.
+	Retries int
+	// V is an instance of Viper.
+	V *viper.Viper
+}
+
+// WithBackoff initializes and returns a heartbeat handle option, which
+// can be used in a heartbeat processing pipeline to prevent trying to send
+// a heartbeat when the api is unresponsive.
+func WithBackoff(config Config) heartbeat.HandleOption {
+	return func(next heartbeat.Handle) heartbeat.Handle {
+		return func(hh []heartbeat.Heartbeat) ([]heartbeat.Result, error) {
+			log.Debugln("execute heartbeat backoff algorithm")
+
+			nextDuration := time.Duration(float64(factor)*math.Pow(2, float64(config.Retries))) * time.Second
+
+			now := time.Now()
+			if config.At != nil && now.Before(config.At.Add(nextDuration)) &&
+				now.Before(config.At.Add(resetAfter*time.Second)) {
+				log.Debugf("current time '%s' is before next backoff at '%s'",
```

Comment:
I guess you mean `backoff end` or `retry` here, right?
```suggestion
				log.Debugf("current time '%s' is before next backoff end at '%s'",
```

Reply:
Maybe only `"current time '%s' is before next backoff"`.

Reply:
I changed this in 89a70b9b0868ed65e9d6eaeb7c0b690233f6509d.